crates/red_knot/src/program/format.rs

```diff
@@ -0,0 +1,44 @@
+use crate::db::{QueryResult, SourceDb};
+use crate::format::{FormatDb, FormatError, FormattedFile};
+use crate::program::Program;
+
+impl Program {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn format(&mut self) -> QueryResult<()> {
+        // Formats all open files
+
+        // TODO make `Executor` from `check` reusable.
+        for file in self.workspace.open_files() {
+            match self.format_file(file) {
+                Ok(FormattedFile::Formatted(content)) => {
+                    let path = self.file_path(file);
+
+                    // TODO: This is problematic because it immediately re-triggers the file watcher.
```

Comment:
It seems like mostly this is a good thing? The file content has changed, the rest of the system should react accordingly, as it normally would. I don't think we should special-case re-formatting in the rest of our invalidation -- it's a file change like any other.

The only place where this seems undesirable is that you would like to track that a file that we just formatted is still formatted; we don't have to reformat it. So I would scope any special-casing to this specific point only, contained within the formatter logic. Maybe we could keep a cache mapping the hash of known-well-formatted contents for a file, so when we get updated sources for a file from the file watcher, we can check if the hash of the new contents matches an entry in this cache, and if so, directly mark it as already well-formatted?